File: cpp/src/ppk/bitio.cc
// Copyright (c) 2009 Dmitri Nikulin
// See LICENSE-MIT.txt for details
// Part of the PPK project

#include "ppk/reader.hh"
#include "ppk/writer.hh"

#include <cassert>

namespace ppk {

static uint64_t kZero64 = 0;
static uint64_t kOne64 = 1;
static uint64_t kFull64 = ~kZero64;

static uint64_t mask64(uint64_t bits) {
    assert (bits <= 64);
    if (bits == 64)
        return kFull64;
    return (kOne64 << bits) - kOne64;
}

uint64_t Reader::getBits(uint8_t bits) {
    assert (m_bit <= 8);
    assert (bits > 0);
    assert (bits <= 64);

    uint64_t out = 0;
    uint8_t done = 0;

    while (done < bits) {
        assert (m_bit <= 8);

        uint8_t more = std::min(8 - m_bit, bits - done);
        assert (more <= 8);

        if (more == 0) {
            m_byte = m_in.get();
            m_bit = 0;
            continue;
        }

        uint64_t value = m_byte >> m_bit;
        value &= mask64(more);
        assert (value <= 0xFF);

        out |= (value << done);
        m_bit += more;
        done += more;
        assert (done <= bits);
        assert (m_bit <= 8);
    }

    assert ((out & mask64(bits)) == out);
    return out;
}

void Reader::skipBits() {
    m_byte = 0;
    m_bit = 8;
}

void Writer::putBits(uint8_t bits, uint64_t value) {
    assert (m_bit < 8);
    assert (bits > 0);
    assert (bits <= 64);

    assert ((value & mask64(bits)) == value);

    uint8_t done = 0;

    while (done < bits) {
        assert (m_bit < 8);

        uint8_t more = std::min(8 - m_bit, bits - done);
        assert (more > 0);
        assert (more <= 8);

        uint64_t downvalue = value >> done;
        assert (downvalue <= value);
        assert ((downvalue & mask64(bits)) == downvalue);

        uint8_t byte = (uint8_t) (downvalue & mask64(more));
        uint8_t shifted = byte << m_bit;
        assert (shifted >= byte);
        assert ((byte == 0) == (shifted == 0));

        m_byte |= shifted;
        m_bit += more;
        done += more;
        assert (done <= bits);
        assert (m_bit <= 8);

        if (m_bit == 8)
            flushBits();
    }
}

void Writer::flushBits() {
    if (m_bit > 0) {
        m_out.put(m_byte);
        m_byte = 0;
        m_bit = 0;
    }
}

}

File: python/ppk/bitio.py
# Copyright (c) 2010 Dmitri Nikulin
# See LICENSE-MIT.txt for details
# Part of the PPK project

'''
Convenience classes for bit stream encoding/decoding.
'''

__all__ = ['BitReader', 'BitWriter', 'mask', 'read0', 'write0']


class BitReader(object):
    '''Bit stream reader.'''

    def __init__(self, getbyte):
        '''
        Create a bit stream reader with the given byte reader function.
        getbyte() must return a Python int 0<=b<=255, or raise EOFError.
        '''
        self.getbyte = getbyte
        self.skip()

    def skip(self):
        '''
        Skip the rest of the byte, if any, buffered from the reader function.
        The next read request will draw at least one byte from the reader function.
        '''
        self.byte = 0
        self.byte_bits = 8

    def __call__(self, bits):
        '''
        Read bits from the buffered byte (if any) and any further bytes (if necessary).
        '''
        assert bits > 0
        assert self.byte_bits >= 0
        assert self.byte_bits <= 8

        out = 0
        done = 0

        while done < bits:
            assert self.byte_bits >= 0
            assert self.byte_bits <= 8

            more = min(8 - self.byte_bits, bits - done)
            assert more >= 0
            assert more <= 8

            if more < 1:
                self.byte = int(self.getbyte())
                assert self.byte >= 0
                assert self.byte <= 0xFF
                self.byte_bits = 0
                continue

            value = self.byte >> self.byte_bits
            value &= mask(more)
            assert value >= 0
            assert value <= 0xFF

            value <<= done
            out |= value

            done += more
            assert done <= bits

            self.byte_bits += more
            assert self.byte_bits >= 0
            assert self.byte_bits <= 8

        assert (out & mask(bits)) == out
        return out



class BitWriter(object):
    '''Bit stream writer.'''

    def __init__(self, putbyte):
        '''
        Create a bit stream writer with the given byte writer function.
        putbyte() must accept a Python int 0<=b<=255.
        '''
        self.putbyte = putbyte
        self.byte = 0
        self.byte_bits = 0

    def flush(self):
        '''
        Write the current byte, if any, to the byte writer function.
        The next bit write request will start a new byte.
        '''
        if self.byte_bits > 0:
            self.byte_bits = 0
            self.putbyte(self.byte)
        self.byte = 0

    def __call__(self, bits, value):
        '''
        Write bits to the stream, flushing bytes as necessary.
        '''
        assert bits > 0
        assert self.byte_bits >= 0
        assert self.byte_bits <= 8

        assert (value & mask(bits)) == value

        done = 0

        while done < bits:
            assert self.byte_bits >= 0
            assert self.byte_bits <= 8

            more = min(8 - self.byte_bits, bits - done)
            assert more > 0
            assert more <= 8

            downvalue = value >> done
            assert downvalue <= value
            assert (downvalue & mask(bits)) == downvalue

            byte = downvalue & mask(more)
            shifted = byte << self.byte_bits
            assert shifted >= byte
            assert (byte == 0) == (shifted == 0)

            self.byte |= shifted

            done += more
            assert done <= bits

            self.byte_bits += more
            assert self.byte_bits >= 0
            assert self.byte_bits <= 8

            if self.byte_bits >= 8:
                self.flush()


def mask(bits):
    '''Calculate an integer mask 'bits' wide'''
    return (1L << bits) - 1L

def read0(*_, **__):
    '''Byte reader mock that always returns 0.'''
    return 0

def write0(*_, **__):
    '''Byte writer mock that ignores all parameters.'''
    pass

def readSequence(seq):
    '''Construct a bit reader that draws bytes from an iterable object.'''
    return BitReader(seq.__iter__().next)

def writeList(out):
    '''Construct a bit writer that appends bytes to a list.'''
    return BitWriter(out.append)
